pulsar4s-akka-streams/src/main/scala/com/sksamuel/pulsar4s/akka/streams/PulsarMultiSinkGraphStage.scala

```diff
@@ -0,0 +1,85 @@
+package com.sksamuel.pulsar4s.akka.streams
+
+import akka.Done
+import akka.stream.stage.{AsyncCallback, GraphStageLogic, GraphStageWithMaterializedValue, InHandler}
+import akka.stream.{Attributes, Inlet, SinkShape}
+import com.sksamuel.exts.Logging
+import com.sksamuel.pulsar4s.{Producer, ProducerMessage, Topic}
+
+import scala.concurrent.{ExecutionContextExecutor, Future, Promise}
+import scala.util.{Failure, Success}
+
+class PulsarMultiSinkGraphStage[T](createFn: Topic => Producer[T], initTopics: Set[Topic] = Set.empty)
+  extends GraphStageWithMaterializedValue[SinkShape[(Topic, ProducerMessage[T])], Future[Done]]
+    with Logging {
+
+  private val in = Inlet.create[(Topic, ProducerMessage[T])]("pulsar.in")
+
+  override def shape: SinkShape[(Topic, ProducerMessage[T])] = SinkShape.of(in)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Done]) = {
+
+    val promise = Promise[Done]()
+
+    val logic: GraphStageLogic = new GraphStageLogic(shape) with InHandler {
+      setHandler(in, this)
+
+      implicit def context: ExecutionContextExecutor = super.materializer.executionContext
+
+      var producers: Map[Topic, Producer[T]] = _
+      var next: AsyncCallback[(Topic, ProducerMessage[T])] = _
+      var error: Throwable = _
+
+      override def preStart(): Unit = {
+        producers = initTopics.map(t => t -> createFn(t)).toMap
+        next = getAsyncCallback { _ => pull(in) }
+        pull(in)
+      }
+
+      private def getProducer(topic: Topic): Producer[T] =
+        producers.getOrElse(topic, {
+          logger.debug(s"creating new producer for topic $topic")
+          val producer = createFn(topic)
+          producers = producers + (topic -> producer)
+          producer
+        })
```

Comment:
I think this could be:
```suggestion
        producers.getOrElseUpdate(topic, {
          logger.debug(s"creating new producer for topic $topic")
          createFn(topic)
        })
```

Reply:
Well, I accepted and then rolled back. We are using immutable map here while keeping it in a `var`. AFAIK `getOrElseUpdate` is available only for mutable maps. We either move to mutable, or keep it as it is. In my opinion it's better to keep it, just seems like immutable map in a `var` is clearer and easier to understand and reason. I always prefer "vars" over mutable collections, but maybe I'm wrong here?

Reply:
Oh, I missed that this was a `val`. I guess it threw me off that you're updating the `producers` in the default value of the `getOrElse`. If we're using the `var`, then this might be clearer:

```suggestion
        producers.get(topic) match {
          case Some(p) => p
          case None =>
            val producer = createFn(topic)
            producers += topic -> producer
            producer
        }
```

That said, I don't see a strong reason to go for an immutable map with a `var` over a mutable map with a `val`. In this context it probably doesn't matter much though.

Reply:
Agree, implemented your suggestion in the newest commit.